scripts/benchmarks_ci.py

```diff
@@ -221,25 +222,37 @@ def __main(args: list) -> int:
 
     # Run micro-benchmarks
     if not args.build_only:
-        for framework in args.frameworks:
-            micro_benchmarks.run(
-                BENCHMARKS_CSPROJ,
-                args.configuration,
-                framework,
-                verbose,
-                args
-            )
-            
-        dotnet.shutdown_server(verbose)
-
-        if args.upload_to_perflab_container:
-            import upload
+        upload_container = UPLOAD_CONTAINER
+        try:
+            for framework in args.frameworks:
+                micro_benchmarks.run(
+                    BENCHMARKS_CSPROJ,
+                    args.configuration,
+                    framework,
+                    verbose,
+                    args
+                )
             globpath = os.path.join(
                 get_artifacts_directory() if not args.bdn_artifacts else args.bdn_artifacts,
                 '**',
                 '*perf-lab-report.json')
+        except CalledProcessError:
+            upload_container = 'failedresults'
+            globpath = os.path.join(
+                get_artifacts_directory(), 
+                'FailureReporter', 
+                'failure-report.json')
+            cmdline = [
+                'FailureReporting.exe', globpath
+            ]
+            RunCommand(cmdline, verbose=verbose).run(
+                os.path.join(get_artifacts_directory()))
```

Comment:
Is this going to work properly in the context of the lab machines? In the other examples the build location is `$(CorrelationStaging)`. In local testing `get_artifacts_directory()` will work out, because the build winds up putting the tool in (mostly) the right place. (I'm actually a bit surprised this didn't need to be `get_artifacts_directory()\FailureReporting`.)

Reply:
Do you mean get_artifacts_directory()\FailureReporter (based on the yaml)?  Is there a way to confirm that that directory would work?

Reply:
If I check out your branch and build FailureReporter, it shows up here:
```
 D:\source\performance\src\tools\Reporting\FailureReporting   upload_failed_runs ≢ +1 ~0 -0 !                                                                                                                                                                                                       [12:11]
❯ dir D:\source\performance\artifacts\bin\FailureReporting\Debug\netcoreapp3.1\

    Directory: D:\source\performance\artifacts\bin\FailureReporting\Debug\netcoreapp3.1

Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
-a---           3/24/2021 12:11 PM          17036 FailureReporting.deps.json
-a---           3/24/2021 12:11 PM          11264 FailureReporting.dll
-a---           3/24/2021 12:11 PM         175104 FailureReporting.exe
-a---           3/24/2021 12:11 PM            180 FailureReporting.runtimeconfig.dev.json
-a---           3/24/2021 12:11 PM            154 FailureReporting.runtimeconfig.json
-a---           5/15/2018  9:01 PM          24024 Microsoft.DotNet.PlatformAbstractions.dll
-a---           4/22/2019  1:06 AM         669608 Newtonsoft.Json.dll
-a---           3/24/2021 12:11 PM          23040 Reporting.dll
```

`get_artifacts_directory()` simply returns `performance\artifacts`:
https://github.com/dotnet/performance/blob/5927768185cd876206f4001c5077964887ee6ce7/scripts/performance/common.py#L88

So I'm surprised your existing code worked, as `FailureReporting.exe` should not exist in the root of the `artifacts` directory.

However, this is a slightly different case than normal. `benchmarks_ci.py` tries hard to work in both local user contexts as well as lab contexts. Some pieces only need to work in the lab, however. This piece of it only ever has to work in the lab, and should never work in local user contexts. The important place to know this landed is somewhere in the correlation payload. By building to `$(CorrelationStaging)`, you ensure that the binary will get picked up and put in the right place by Helix. To find that on the other end, you can use a method like we have in the scenarios code:
https://github.com/dotnet/performance/blob/09356faeb90bbfefa0a9f128b63208ad5f95735f/src/scenarios/shared/util.py#L11-L15

I would not reference that file (`util.py`) directly from here, so perhaps some refactoring is useful to move that method to `common.py` or something (and then fix up existing call sites.)